Improve work package

brycemcanally opened the pull request:
This PR adds a task queue to the work package and finishes some of the todos around canceling and cleaning up tasks. This PR also changes the layout of work in etcd and includes some changes that should reduce the number of etcd requests. The basic way to interface with the task queue for setting up parallel tasks is:
```
tq := work.NewTaskQueue(ctx, etcdClient, etcdPrefix, taskNameSpace)
...
// An alternative is tq.RunTaskBlock(ctx, func(m *Master) error {...})
if err := tq.RunTask(ctx, func(m *Master) {
  // An alternative is m.RunSubtasksChan(subtaskChan, collectFunc)
  if err := m.RunSubtasks(subtasks, collectFunc); err != nil {
    return err
  }
  ...
  if err := m.RunSubtasks(otherSubtasks, otherCollectFunc); err != nil {
    return err
  }
  ...
}); err != nil {
  // log
}
```
The worker interface remains the same. Priority in the task queue is determined by the order in which the tasks are created. Tasks and subtasks are scoped to the corresponding `Run*` functions, so the etcd entries will be cleaned upon return.

This PR also integrates these changes into the compaction process. A few more changes to the compaction process were necessary to take full advantage of these changes (and work correctly).

Tryneus commented:
Before I do a deep-dive into the code, I wonder if the usage could be simplified:

```golang
if err := func() error {
    defer func() {
        if err := tq.DeleteTask(task.ID); retErr == nil {
            retErr = err
        }
    }()
    m.Collect(collectFunc)
    for _, subtask := range subtasks {
        if err := m.CreateSubtask(subtask); err != nil {
            return err
        }
    }
    return m.Wait()
}(); err != nil {
    // log error
}
```
This is only really doing two things - providing subtasks and collecting results, the rest is just boilerplate.  The boilerplate could be moved into the task queue itself (even if it's just porcelain over the same underlying calls):

```golang
// subtaskChan := make(chan work.Subtask)
if err := m.RunSubtasks(subtaskChan, collectFunc); err != nil {
  // log error
}
```

I would also argue that (at least for the worker package), the `DeleteTask` would not be performed in that IIFE (or in `RunSubtasks` in my example) - we rely on the lifespan of the work master to guarantee our spot in line in case more tasks are added in the future (i.e. we call `RunSubtasks` once for processing datums and once for merging).

I'd prefer something like a `work.WithMaster` call that scopes the lifetime of the master, but that makes it harder to instantiate them in correct order, given that this requires `work.WithMaster` to be called in a separate goroutine from the spawner goroutine.  Not that I can't do some `sync` magic to line it back up, it just doesn't simplify much.  As it stands, I dislike having cleanup tasks be enforced by the user, as it's easy to forget to do it or to miss a contingency and leak resources, so I'd prefer this is done by the `work` package explicitly.

brycemcanally commented:
Yeah, so this is what I was alluding to with the last sentence in my original post. My original goal with this was to have the api such that you could specify a callback in the create task function which would put both the cleanup and the parallelism management within the work package. That does not really work with your job parallelism stuff because you have two different task types in a job (although we could probably compose the callbacks, but I think that might be kind of weird). I think when we move to the new storage layer, we might be able to map a single task in the work package to a single job which would allow us to manage both the job parallelism and cleanup in the work package in a clean and simple way.

> I would also argue that (at least for the worker package), the DeleteTask would not be performed in that IIFE (or in RunSubtasks in my example) - we rely on the lifespan of the work master to guarantee our spot in line in case more tasks are added in the future (i.e. we call RunSubtasks once for processing datums and once for merging).

Not sure I understand what you are getting at here. There would be two task queues (one for processing and one for merging). Both tasks (processing/merging) would get setup outside of the goroutine (which corresponds to the job goroutine in your job parallelism stuff). When the job completes (job goroutine returns), then delete task would be called for each task. The example is just for one task queue.

```
// subtaskChan := make(chan work.Subtask)
if err := m.RunSubtasks(subtaskChan, collectFunc); err != nil {
  // log error
}
```
This just seems like unnecessary concurrency. What benefit do we get from operating on a channel here (in terms of the work package in general, not just how you currently have things coded)? The implementation of making a call per subtask (or even having an api for a batch) is much simpler and performant enough.

Tryneus commented on src/server/pkg/work/testing/work_test.go:
```diff
@@ -0,0 +1,151 @@
+			go func() {
+				for {
+					ctx, cancel := context.WithCancel(context.Background())
+					w := work.NewWorker(context.Background(), env.EtcdClient, etcdPrefix, "")
```
It appears that this is necessary for the test to pass (`TestWorkerCrashes` hangs otherwise).  I'm not sure why, but I would guess that canceling the context leaves the worker in a bad state.

brycemcanally replied:
This is intended to simulate the worker machine crashing. Calling run multiple times might be an issue with the queuing (since it is setup in the `NewWorker` call). I might need to make some changes here.

Tryneus commented on src/server/pkg/work/task_queue.go:
```diff
@@ -0,0 +1,104 @@
+		for {
+			select {
+			case <-ctx.Done():
+				fmt.Printf("task queue context canceled: %v\n", ctx.Err())
```
This doesn't actually do anything about it - this will loop indefinitely, printing this out.

brycemcanally replied:
Fixed.

Tryneus commented on src/server/pkg/work/work.go:
```diff
@@ -83,140 +100,196 @@ func (m *Master) Run(ctx context.Context, task *Task) (retErr error) {
+		select {
+		case e := <-claimWatch.Watch():
+			if e.Type == watch.EventError {
 				return err
```
I think this (and line 235) should be `return e.Err`

brycemcanally replied:
Fixed.

brycemcanally commented:
> I don't see why I would want to do it this way instead of differentiating processing and merging tasks within the same queue. Otherwise I'll need separate concurrency controls to make sure a worker is only processing one task across two queues (which could get fairly complicated).

I don't think this would be worth the complexity right now when the merge process is going to be moved out of PPS. Why can't a processing and merge subtask run in parallel?

> I suppose I don't need the work package to provide an easy to use interface, since I can just wrap that API in something more reliable, but I don't see a use case for having separate calls for these things.

> After looking at the code, I feel like this is way more complicated than it needs to be, and it's very hard to reason about concurrency and corner-cases (but maybe it's just me). I'm going to take a stab at this myself to see if I'm crazy.

My thinking on this is that it should be temporary boilerplate until the new storage layer is released. It does not seem like you really addressed the path forward I had in mind (map a single task to a single job after the merge process is moved out of PPS), but I am guessing you do not agree with it? If you feel like there is a better way to do this, I am open to it, but I would at least like some insight into what you have in mind. I will hold off on addressing the feedback until you tell me what changes you want to make and why.

brycemcanally commented on src/server/pkg/work/work.go:
```diff
@@ -83,140 +100,196 @@ func (m *Master) Run(ctx context.Context, task *Task) (retErr error) {
+	// Cleanup subtask entries.
+	_, err := col.NewSTM(ctx, tq.etcdClient, func(stm col.STM) error {
+		tq.subtaskCol.ReadWrite(stm).DeleteAllPrefix(taskID)
+		return nil
```
No, they shouldn't. Fixed.

brycemcanally commented on src/server/pkg/work/work.go:
```diff
@@ -83,140 +100,196 @@ func (m *Master) Run(ctx context.Context, task *Task) (retErr error) {
 // The worker will continue to watch the task collection until the context is cancelled.
-func (w *Worker) Run(ctx context.Context) error {
+func (w *Worker) Run(ctx context.Context, processFunc ProcessFunc) error {
 	return w.taskCol.ReadOnly(ctx).WatchF(func(e *watch.Event) (retErr error) {
```
Yes, our watch package sorts by mod revision for the initial get.

brycemcanally commented on src/server/pkg/work/task_queue.go:
```diff
@@ -0,0 +1,104 @@
+				}
+			}
+			tq.mu.Unlock()
+			time.Sleep(waitTime)
```
This was discussed offline. Polling is a simple approach that will work for our needs (although we don't really like it).

brycemcanally commented on src/server/pkg/work/testing/work_test.go:
```diff
@@ -0,0 +1,151 @@
+							processMu.Lock()
+							defer processMu.Unlock()
+							if subtasksProcessed[i] != nil {
+								require.False(t, subtasksProcessed[i][subtask.ID], msg)
```
Okay, I can change that.

brycemcanally commented on src/server/pkg/work/testing/work_test.go:
```diff
@@ -0,0 +1,151 @@
+						if ctx.Err() == context.Canceled {
+							continue
+						}
+						break
```
Yep, fixed.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/pachyderm/pachyderm?pullRequest=4650) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/pachyderm/pachyderm?pullRequest=4650) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/pachyderm/pachyderm?pullRequest=4650) it.</sub>

Tryneus commented on src/server/pkg/work/work.go:
```diff
@@ -117,106 +199,124 @@ type ProcessFunc func(context.Context, *Task, *Task) error
+			if err := e.Unmarshal(&subtaskKey, &TaskInfo{}); err != nil {
+				return err
+			}
+			if err := taskEntry.runSubtask(w.subtaskFunc(task, subtaskKey, processFunc)); err != nil {
```
So I think you should be able to skip this part if the subtask state isn't `RUNNING`.  I tried this with my benchmarks and it got slower, though, which is pretty weird.

brycemcanally replied:
Yeah, that is true, although I don't think it really matters with the queuing model. You are basically avoiding two channel writes and an iteration through the tasks in the queue by doing this. I can still add this, but I question if it is worth the extra lines of code considering the task queue is not expected to have a very large number of tasks (so I would like to keep it as simple as possible). It does seem weird that this made your benchmark slower. I guess it could technically result with more etcd contention because you could have the subtask that is running and events from the claim watcher performing requests in parallel.

Tryneus commented on src/server/pkg/work/work.go:
```diff
@@ -117,106 +199,124 @@ type ProcessFunc func(context.Context, *Task, *Task) error
+			return w.subtaskCol.ReadWrite(stm).Get(subtaskKey, subtaskInfo)
+		}); err != nil {
+			return err
+		}
```
Could you put the read inside the `Claim` block?  Or is it common for workers to claim tasks that are already completed?

brycemcanally replied:
If you notice the comment above this, I would like to refactor this so that the read and claim happen in the same stm. Although this is not really critical, it does prevent a rare race condition that does not affect correctness, but could result with duplicate work (although I prevent updating the subtask entry when this occurs). Basically, we could have a worker see that a subtask is running, then attempt to claim it immediately after a different worker has finished it and the claim expired (so the original worker thinks it has not been processed and does not see a claim on it). I added a check after a subtask completes to see if the subtask is already completed, which will result with the worker that did the duplicate work to not update the entry (we want the results from when it was originally processed).

Tryneus commented on src/server/pkg/work/work.go:
```diff
@@ -14,66 +17,117 @@ import (
+			if err := tq.deleteTask(task.ID); err != nil {
+				fmt.Printf("errored deleting task %v: %v\n", task.ID, err)
+			}
+		}()
```
It looks like `taskQueue.runTask` already cleans up the task before returning - I _think_ both the defers in this function can go away.

brycemcanally replied:
I think all three deletes are necessary. The one in `runTask` removes the entry in the `taskQueue` (internal data structure that actually implements the queuing behavior). The two in `RunTask` are necessary for cleaning up the etcd state associated with `RunTask`. There is one that handles the case where we error when attempting to setup the task in the internal task queue, but after setting up the etcd entry (The etcd entry for the task needs to be correctly setup before we can run the callback passed in by the user). The other is the happy path for cleanup after the client callback is completed.

Tryneus replied:
Ah, I see - I was confused between `taskQueue.deleteTask` and `TaskQueue.deleteTask` - I really think one of these should be renamed...

Tryneus commented on src/server/pkg/work/work.go:
```diff
@@ -83,140 +147,203 @@ func (m *Master) Run(ctx context.Context, task *Task) (retErr error) {
+					subtaskInfo := &TaskInfo{}
+					if _, err := col.NewSTM(claimCtx, w.etcdClient, func(stm col.STM) error {
+						return w.subtaskCol.ReadWrite(stm).Update(subtaskKey, subtaskInfo, func() error {
+							// (bryce) remove when check and claim are in the same stm.
```
Something that came up during integration with my branch - this update no longer saves the latest `subtaskInfo.Task.Data` field upon completion of the task, which was behavior I was using to return results to the master.

brycemcanally replied:
Yeah, that is the behavior we want. Working on a fix now.

Tryneus commented on src/server/pkg/work/task_queue.go:
```diff
@@ -0,0 +1,146 @@
+// remap copies the entries in the ordered map to a new ordered map after a certain number of
+// tasks have been encountered. This is to prevent unbounded memory usage due to maps not freeing
+// memory after deletions.
+func (tq *taskQueue) remap() {
```
nit - I'd rename this `maybeRemap` because it doesn't always do the thing

brycemcanally replied:
Done.

Tryneus commented on src/server/pkg/work/task_queue.go:
```diff
@@ -0,0 +1,146 @@
+	if _, ok := tq.tasks.Get(taskID); ok {
+		return fmt.Errorf("errored creating task %v, which already exists", taskID)
+	}
+	tq.remap()
```
I think you should move this to when tasks are removed from the map - if the problem is freeing on empty entries, this seems like the wrong place to do it.

brycemcanally replied:
Makes sense. Done.

jdoliner commented on src/server/pfs/server/driver_new_storage.go:
```diff
@@ -45,45 +44,53 @@ func (d *driver) finishCommitNewStorageLayer(txnCtx *txnenv.TransactionContext,
+		// (bryce) need some cleanup improvements, probably garbage collection.
+		// (bryce) this should be a retry when garbage collection is integrated.
+		d.storage.Delete(context.Background(), path.Join(commitPath, fileset.Diff))
+		d.storage.Delete(context.Background(), path.Join(commitPath, fileset.Compacted))
```
Why are these using `context.Background()` when the above uses `txnCtx.Client.Ctx()`?

brycemcanally replied:
This is a hack to get compactions working before garbage collection is integrated. Basically, we need to clean up the diff and compacted file set for failed compactions. This will be handled properly when garbage collection is integrated.

jdoliner commented on src/server/pfs/server/driver_new_storage.go:
```diff
@@ -97,54 +104,43 @@ func (d *driver) getFilesNewStorageLayer(ctx context.Context, repo, commit, glob
+	scratch := path.Join(tmpPrefix, uuid.NewWithoutDashes())
+	defer func() {
+		// (bryce) need some cleanup improvements, probably garbage collection.
+		if err := d.storage.Delete(context.Background(), scratch); retErr == nil {
```
`context.Background` here as well, should probably be a more meaningful context.

brycemcanally replied:
We still want the cleanup code to run if the context is cancelled. But, this again will be handled by garbage collection, so I don't think it is particularly important for getting this PR in (a lot of this integration work in general is going to require some temporary hacks so that we can get this stuff in digestible pieces, but I do make notes of all these places as you can see).

jdoliner commented on src/server/pfs/server/driver_new_storage.go:
```diff
@@ -158,35 +154,31 @@ func serializeShard(shard *pfs.Shard) (*types.Any, error) {
-	w := work.NewWorker(d.etcdClient, d.prefix, func(ctx context.Context, task, subtask *work.Task) (retErr error) {
-		compaction, shard, err := deserialize(task.Data, subtask.Data)
+	w := work.NewWorker(d.etcdClient, d.prefix, storageTaskNamespace)
+	if err := w.Run(context.Background(), func(ctx context.Context, subtask *work.Task) error {
```
`context.Background`

brycemcanally replied:
This is the correct context for a PFS compaction worker. It should always be running.
